File: android/app/src/main/java/com/xax/CryptoSavingsTracker/domain/navigation/NavigationTelemetry.kt
package com.xax.CryptoSavingsTracker.domain.navigation

import android.util.Log
import javax.inject.Inject
import javax.inject.Singleton

enum class NavigationTelemetryEvent(val wireName: String) {
    FLOW_STARTED("nav_flow_started"),
    FLOW_COMPLETED("nav_flow_completed"),
    CANCELLED("nav_cancelled"),
    DISCARD_CONFIRMED("nav_discard_confirmed"),
    RECOVERY_COMPLETED("nav_recovery_completed")
}

object NavigationJourney {
    const val GOAL_CREATE_EDIT = "goal-create-edit"
    const val MONTHLY_BUDGET_ADJUST = "monthly-budget-adjust"
    const val DESTRUCTIVE_DELETE_CONFIRMATION = "destructive-delete-confirmation"
    const val GOAL_CONTRIBUTION_EDIT_CANCEL = "goal-contribution-edit-cancel"
    const val PLANNING_FLOW_CANCEL_RECOVERY = "planning-flow-cancel-recovery"

    val TOP_5: Set<String> = setOf(
        GOAL_CREATE_EDIT,
        MONTHLY_BUDGET_ADJUST,
        DESTRUCTIVE_DELETE_CONFIRMATION,
        GOAL_CONTRIBUTION_EDIT_CANCEL,
        PLANNING_FLOW_CANCEL_RECOVERY
    )
}

data class NavigationTelemetryPayload(
    val event: NavigationTelemetryEvent,
    val journeyId: String,
    val platform: String,
    val entryPoint: String? = null,
    val durationMs: Long? = null,
    val result: String? = null,
    val isDirty: Boolean? = null,
    val cancelStage: String? = null,
    val formType: String? = null,
    val recoveryPath: String? = null,
    val success: Boolean? = null
) {
    fun properties(): Map<String, String> {
        val values = linkedMapOf(
            "journey_id" to journeyId,
            "platform" to platform
        )
        entryPoint?.let { values["entry_point"] = it }
        durationMs?.let { values["duration_ms"] = it.toString() }
        result?.let { values["result"] = it }
        isDirty?.let { values["is_dirty"] = it.toString() }
        cancelStage?.let { values["cancel_stage"] = it }
        formType?.let { values["form_type"] = it }
        recoveryPath?.let { values["recovery_path"] = it }
        success?.let { values["success"] = it.toString() }
        return values
    }

    fun requiredFieldViolations(): List<String> {
        val required = when (event) {
            NavigationTelemetryEvent.FLOW_STARTED -> listOf("journey_id", "platform", "entry_point")
            NavigationTelemetryEvent.FLOW_COMPLETED -> listOf("journey_id", "platform", "duration_ms", "result")
            NavigationTelemetryEvent.CANCELLED -> listOf("journey_id", "platform", "is_dirty", "cancel_stage")
            NavigationTelemetryEvent.DISCARD_CONFIRMED -> listOf("journey_id", "platform", "form_type")
            NavigationTelemetryEvent.RECOVERY_COMPLETED -> listOf("journey_id", "platform", "recovery_path", "success")
        }
        val props = properties()
        return required.filter { key -> props[key].isNullOrBlank() }
    }
}

interface NavigationTelemetryProvider {
    fun track(payload: NavigationTelemetryPayload)
}

@Singleton
class LogcatNavigationTelemetryProvider @Inject constructor() : NavigationTelemetryProvider {
    override fun track(payload: NavigationTelemetryPayload) {
        val serialized = payload.properties()
            .toSortedMap()
            .entries
            .joinToString(separator = ",") { "${it.key}=${it.value}" }

        val violations = payload.requiredFieldViolations()
        if (violations.isNotEmpty()) {
            Log.w(
                "NavigationTelemetry",
                "[${payload.event.wireName}] schema_violation=${violations.joinToString("|")} $serialized"
            )
        }

        Log.i("NavigationTelemetry", "[${payload.event.wireName}] $serialized")
    }
}

@Singleton
class NavigationTelemetryTracker @Inject constructor(
    private val provider: NavigationTelemetryProvider
) {
    private val flowStartTimes: MutableMap<String, Long> = mutableMapOf()
    private val lastEventByFingerprint: MutableMap<String, Long> = mutableMapOf()

    var dedupeWindowMs: Long = 800L
    internal var clock: () -> Long = { System.currentTimeMillis() }

    fun flowStarted(journeyId: String, entryPoint: String) {
        flowStartTimes[journeyId] = clock()
        emit(
            NavigationTelemetryPayload(
                event = NavigationTelemetryEvent.FLOW_STARTED,
                journeyId = journeyId,
                platform = "android",
                entryPoint = entryPoint
            )
        )
    }

    fun flowCompleted(journeyId: String, result: String = "success") {
        val now = clock()
        val startedAt = flowStartTimes.remove(journeyId) ?: now
        emit(
            NavigationTelemetryPayload(
                event = NavigationTelemetryEvent.FLOW_COMPLETED,
                journeyId = journeyId,
                platform = "android",
                durationMs = (now - startedAt).coerceAtLeast(0L),
                result = result
            )
        )
    }

    fun cancelled(journeyId: String, isDirty: Boolean, cancelStage: String) {
        emit(
            NavigationTelemetryPayload(
                event = NavigationTelemetryEvent.CANCELLED,
                journeyId = journeyId,
                platform = "android",
                isDirty = isDirty,
                cancelStage = cancelStage
            )
        )
    }

    fun discardConfirmed(journeyId: String, formType: String) {
        emit(
            NavigationTelemetryPayload(
                event = NavigationTelemetryEvent.DISCARD_CONFIRMED,
                journeyId = journeyId,
                platform = "android",
                formType = formType
            )
        )
    }

    fun recoveryCompleted(journeyId: String, recoveryPath: String, success: Boolean) {
        emit(
            NavigationTelemetryPayload(
                event = NavigationTelemetryEvent.RECOVERY_COMPLETED,
                journeyId = journeyId,
                platform = "android",
                recoveryPath = recoveryPath,
                success = success
            )
        )
    }

    private fun emit(payload: NavigationTelemetryPayload) {
        val now = clock()
        val fingerprint = buildString {
            append(payload.event.wireName)
            payload.properties()
                .toSortedMap()
                .forEach { (key, value) ->
                    append('|').append(key).append('=').append(value)
                }
        }

        val previous = lastEventByFingerprint[fingerprint]
        if (previous != null && (now - previous) < dedupeWindowMs) {
            return
        }

        lastEventByFingerprint[fingerprint] = now
        provider.track(payload)
    }
}

File: ios/CryptoSavingsTracker/Utilities/NavigationTelemetry.swift
import Foundation

enum NavigationTelemetryEvent: String, CaseIterable {
    case flowStarted = "nav_flow_started"
    case flowCompleted = "nav_flow_completed"
    case cancelled = "nav_cancelled"
    case discardConfirmed = "nav_discard_confirmed"
    case recoveryCompleted = "nav_recovery_completed"
    case goalDashboardOpened = "goal_dashboard_opened"
    case goalDashboardPrimaryCtaShown = "goal_dashboard_primary_cta_shown"
    case goalDashboardPrimaryCtaTapped = "goal_dashboard_primary_cta_tapped"
}

enum NavigationJourney {
    static let goalCreateEdit = "goal-create-edit"
    static let monthlyBudgetAdjust = "monthly-budget-adjust"
    static let destructiveDeleteConfirmation = "destructive-delete-confirmation"
    static let goalContributionEditCancel = "goal-contribution-edit-cancel"
    static let planningFlowCancelRecovery = "planning-flow-cancel-recovery"
    static let goalDashboard = "goal-dashboard"

    static let top5: [String] = [
        goalCreateEdit,
        monthlyBudgetAdjust,
        destructiveDeleteConfirmation,
        goalContributionEditCancel,
        planningFlowCancelRecovery
    ]
}

struct NavigationTelemetryPayload {
    let event: NavigationTelemetryEvent
    let journeyID: String
    let platform: String
    let entryPoint: String?
    let durationMs: Int?
    let result: String?
    let isDirty: Bool?
    let cancelStage: String?
    let formType: String?
    let recoveryPath: String?
    let success: Bool?
    let goalID: String?
    let resolverState: String?
    let ctaID: String?

    init(
        event: NavigationTelemetryEvent,
        journeyID: String,
        platform: String,
        entryPoint: String?,
        durationMs: Int?,
        result: String?,
        isDirty: Bool?,
        cancelStage: String?,
        formType: String?,
        recoveryPath: String?,
        success: Bool?,
        goalID: String? = nil,
        resolverState: String? = nil,
        ctaID: String? = nil
    ) {
        self.event = event
        self.journeyID = journeyID
        self.platform = platform
        self.entryPoint = entryPoint
        self.durationMs = durationMs
        self.result = result
        self.isDirty = isDirty
        self.cancelStage = cancelStage
        self.formType = formType
        self.recoveryPath = recoveryPath
        self.success = success
        self.goalID = goalID
        self.resolverState = resolverState
        self.ctaID = ctaID
    }

    var properties: [String: String] {
        var values: [String: String] = [
            "journey_id": journeyID,
            "platform": platform
        ]

        if let entryPoint { values["entry_point"] = entryPoint }
        if let durationMs { values["duration_ms"] = String(durationMs) }
        if let result { values["result"] = result }
        if let isDirty { values["is_dirty"] = isDirty ? "true" : "false" }
        if let cancelStage { values["cancel_stage"] = cancelStage }
        if let formType { values["form_type"] = formType }
        if let recoveryPath { values["recovery_path"] = recoveryPath }
        if let success { values["success"] = success ? "true" : "false" }
        if let goalID { values["goal_id"] = goalID }
        if let resolverState { values["resolver_state"] = resolverState }
        if let ctaID { values["cta_id"] = ctaID }

        return values
    }

    func requiredFieldViolations() -> [String] {
        switch event {
        case .flowStarted:
            return missing(["journey_id", "platform", "entry_point"])
        case .flowCompleted:
            return missing(["journey_id", "platform", "duration_ms", "result"])
        case .cancelled:
            return missing(["journey_id", "platform", "is_dirty", "cancel_stage"])
        case .discardConfirmed:
            return missing(["journey_id", "platform", "form_type"])
        case .recoveryCompleted:
            return missing(["journey_id", "platform", "recovery_path", "success"])
        case .goalDashboardOpened:
            return missing(["journey_id", "platform", "goal_id", "entry_point"])
        case .goalDashboardPrimaryCtaShown, .goalDashboardPrimaryCtaTapped:
            return missing(["journey_id", "platform", "goal_id", "resolver_state", "cta_id"])
        }
    }

    private func missing(_ required: [String]) -> [String] {
        let values = properties
        return required.filter { key in
            guard let value = values[key] else {
                return true
            }
            return value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }
}

protocol NavigationTelemetryProvider {
    func track(_ payload: NavigationTelemetryPayload)
}

struct AppLogNavigationTelemetryProvider: NavigationTelemetryProvider {
    func track(_ payload: NavigationTelemetryPayload) {
        let ordered = payload.properties
            .sorted(by: { $0.key < $1.key })
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: ",")

        let violations = payload.requiredFieldViolations()
        if !violations.isEmpty {
            AppLog.warning(
                "[\(payload.event.rawValue)] schema_violation=\(violations.joined(separator: "|")) \(ordered)",
                category: .ui
            )
        }

        AppLog.info("[\(payload.event.rawValue)] \(ordered)", category: .ui)
    }
}

@MainActor
final class NavigationTelemetryTracker {
    private let provider: NavigationTelemetryProvider
    private let clock: () -> Date
    private let dedupeWindowMs: Int

    private var flowStartDates: [String: Date] = [:]
    private var lastEventByFingerprint: [String: Date] = [:]

    init(
        provider: NavigationTelemetryProvider,
        clock: @escaping () -> Date = { Date() },
        dedupeWindowMs: Int = 800
    ) {
        self.provider = provider
        self.clock = clock
        self.dedupeWindowMs = dedupeWindowMs
    }

    func flowStarted(journeyID: String, entryPoint: String) {
        flowStartDates[journeyID] = clock()
        emit(
            NavigationTelemetryPayload(
                event: .flowStarted,
                journeyID: journeyID,
                platform: "ios",
                entryPoint: entryPoint,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: nil,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func flowCompleted(journeyID: String, result: String = "success") {
        let now = clock()
        let startedAt = flowStartDates[journeyID] ?? now
        let durationMs = max(0, Int(now.timeIntervalSince(startedAt) * 1000))
        flowStartDates.removeValue(forKey: journeyID)

        emit(
            NavigationTelemetryPayload(
                event: .flowCompleted,
                journeyID: journeyID,
                platform: "ios",
                entryPoint: nil,
                durationMs: durationMs,
                result: result,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: nil,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func cancelled(journeyID: String, isDirty: Bool, cancelStage: String) {
        emit(
            NavigationTelemetryPayload(
                event: .cancelled,
                journeyID: journeyID,
                platform: "ios",
                entryPoint: nil,
                durationMs: nil,
                result: nil,
                isDirty: isDirty,
                cancelStage: cancelStage,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: nil,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func discardConfirmed(journeyID: String, formType: String) {
        emit(
            NavigationTelemetryPayload(
                event: .discardConfirmed,
                journeyID: journeyID,
                platform: "ios",
                entryPoint: nil,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: formType,
                recoveryPath: nil,
                success: nil,
                goalID: nil,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func recoveryCompleted(journeyID: String, recoveryPath: String, success: Bool) {
        emit(
            NavigationTelemetryPayload(
                event: .recoveryCompleted,
                journeyID: journeyID,
                platform: "ios",
                entryPoint: nil,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: recoveryPath,
                success: success,
                goalID: nil,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func goalDashboardOpened(goalID: String, entryPoint: String) {
        emit(
            NavigationTelemetryPayload(
                event: .goalDashboardOpened,
                journeyID: NavigationJourney.goalDashboard,
                platform: "ios",
                entryPoint: entryPoint,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: goalID,
                resolverState: nil,
                ctaID: nil
            )
        )
    }

    func goalDashboardPrimaryCtaShown(goalID: String, resolverState: String, ctaID: String) {
        emit(
            NavigationTelemetryPayload(
                event: .goalDashboardPrimaryCtaShown,
                journeyID: NavigationJourney.goalDashboard,
                platform: "ios",
                entryPoint: nil,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: goalID,
                resolverState: resolverState,
                ctaID: ctaID
            )
        )
    }

    func goalDashboardPrimaryCtaTapped(goalID: String, resolverState: String, ctaID: String) {
        emit(
            NavigationTelemetryPayload(
                event: .goalDashboardPrimaryCtaTapped,
                journeyID: NavigationJourney.goalDashboard,
                platform: "ios",
                entryPoint: nil,
                durationMs: nil,
                result: nil,
                isDirty: nil,
                cancelStage: nil,
                formType: nil,
                recoveryPath: nil,
                success: nil,
                goalID: goalID,
                resolverState: resolverState,
                ctaID: ctaID
            )
        )
    }

    private func emit(_ payload: NavigationTelemetryPayload) {
        let fingerprint = [payload.event.rawValue] + payload.properties.sorted(by: { $0.key < $1.key }).map { "\($0.key)=\($0.value)" }
        let key = fingerprint.joined(separator: "|")
        let now = clock()

        if let last = lastEventByFingerprint[key] {
            let deltaMs = Int(now.timeIntervalSince(last) * 1000)
            if deltaMs < dedupeWindowMs {
                return
            }
        }

        lastEventByFingerprint[key] = now
        provider.track(payload)
    }
}
